<html><head><title>GeneDesign Guide: Version History</title><meta http-equiv="Content-Type" content="text/html; charset=iso-8859-1"><link href="../acss/mn.css" rel="stylesheet" type="text/css"></head><body><div id="bigbox">	<div id="header">		<a href="../index.html"><img src="../img/gdlogobanner.gif" alt="GeneDesign"></a>  <a class="headli">Manual</a>	</div>	<div id="navbox">		<ul id="navlist">			<li><a href="about.html">About GeneDesign</a></li>			<li><a href="gendes.html">Design a Gene</a></li>			<ul id="subnavlist">				<li><a href="dgrevtrans.html">Reverse Translation</a></li>				<li><a href="dgssi1.html">Site Insertion (1)</a></li>				<li><a href="dgssi2.html">Site Insertion (2)</a></li>				<li><a href="dgolides.html">Oligo Design</a></li>			</ul>			<li><a href="mods.html">GeneDesign Modules</a></li>			<ul id="subnavlist">				<li><a href="codjug.html">Codon Juggling</a></li>				<li><a href="revtrans.html">Reverse Translation</a></li>				<li><a href="ssi.html">Silent Site Insertion</a></li>				<li><a href="ssr.html">Silent Site Removal</a></li>				<li><a href="enzcho.html">Enzyme Choosing</a></li>				<li><a href="seqana.html">Sequence Analysis</a></li>				<li><a href="olides.html">Oligo Design</a></li>				<li><a href="veccho.html">Vector Choosing</a></li>				<li><a href="shortr.html">Short Sequence Removal</a></li>				<li><a href="randna.html">Random DNA Sequences</a></li>			</ul>			<li><a href="errrep.html">Report an Error</a></li>			<li><a href="sugcom.html">Suggestions &amp; Comments</a></li>			<li id="active"><a href="enzymes.html">Enzyme List</a></li>			<li id="active"><a href="changes.html" id="current">Version History</a></li>		</ul>	</div><div id="content"><strong>Changes from 1.9 to 2.0</strong><ul>	<li>Revamped Enzyme List and Enzyme Selection Criteria</li>	<ul>		<li>New list composed of prototypes (and their neoschizomers) that are commercially available, single cutting, Type IIs with known cleavage sites.</li>		<li>No longer use buffer, vendor, and NEB availability as criteria.</li>		<li>Added incubation temperature, methylation sensitivity, star activity and exhaustive vendor list.</li>		<li>Now able to exclude enzymes with cleavage external to recognition site.</li>	</ul>	<li>Added C. elegans to list of known organisms</li>	<li>Oligo Design now automatically excludes blunts, 1bp overhangs, and enzymes that cut outside of recognition site.</li>	<li>Codon Juggling detailed</li>	<ul>		<li>Now warns if you are not using a simple coding sequence</li>		<li>Will allow you to not define an organism; hides optimization algorithms and randomizes most different</li>		<li>Define next most optimal and random as highly experimental algorithms</li>	</ul>	<li>Silent Site Insertion detailed</li>	<ul>		<li>Allow reconsideration of automatic site selection</li>		<li>Fixed error messages about simple coding sequences</li>	</ul>	<li>Miscellaneous bug fixes that you probably never noticed</li>	<li>Manual updated accordingly and PDF version added.</li>	<li>Now with version history!</li></ul></div>  <div id="footer">	<a href="about.html" class="headli2">Back</a>  </div></div></body></html>