<html><head><title>GeneDesign Manual: Site Addition (2)</title><meta http-equiv="Content-Type" content="text/html; charset=iso-8859-1"><link href="../acss/mn.css" rel="stylesheet" type="text/css"></head><body><div id="bigbox">	<div id="header">		<a href="../index.html"><img src="../img/gdlogobanner.gif" alt="GeneDesign"></a>  <a class="headli">Manual</a>	</div>	<div id="navbox">		<ul id="navlist">			<li><a href="about.html">About GeneDesign</a></li>			<li id="active"><a href="gendes.html" id="current">Design a Gene</a></li>			<ul id="subnavlist">				<li><a href="dgrevtrans.html">Reverse Translation</a></li>				<li><a href="dgssi1.html">Site Addition (1)</a></li>				<li id="active"><a href="dgssi2.html" id="current">Site Addition (2)</a></li>				<li><a href="dgolides.html">Building Block Design</a></li>			</ul>			<li><a href="mods.html">GeneDesign Modules</a></li>			<ul id="subnavlist">				<li><a href="codjug.html">Codon Juggling</a></li>				<li><a href="revtrans.html">Reverse Translation</a></li>				<li><a href="codbigr.html">Codon Bias Graphing</a></li>				<li><a href="rscugen.html">RSCU Value Generation</a></li>				<li><a href="ssi.html">Restriction Site Addition</a></li>				<li><a href="ssr.html">Restriction Site Subtraction</a></li>				<li><a href="enzcho.html">Enzyme Filter</a></li>				<li><a href="seqana.html">Sequence Analysis</a></li>				<li><a href="olides.html">Building Block Design</a></li>				<li><a href="veccho.html">Vector Choosing</a></li>				<li><a href="shortr.html">Short Sequence Subtraction</a></li>				<li><a href="randna.html">Random DNA Sequences</a></li>			</ul>			<li><a href="errrep.html">Report an Error</a></li>			<li><a href="sugcom.html">Suggestions &amp; Comments</a></li>			<li><a href="enzymes.html">Enzyme List</a></li>			<li><a href="changes.html">Version History</a></li>		</ul>	</div>	<div id="content">  <strong>Design a Gene: Restriction Site Addition, part 2</strong><br>  <br>  The second step of Restriction Site Addition involves the actual   selection of landmark restriction sites. Your protein sequence is displayed with an asterisk under every  tenth amino acid.  Anchored pull down menus contain all pf the possible silent, unique insertions.<br>  <br><div id="imgrt">  <img src="../img/ssi2.gif" alt="Restriction Site Addition Module" width="300" height="251"><br>  The Restriction Site Addition Module - Landmark Selection</div>  <p>    Easiest method: You can make the computer select sites for you with the &quot;Pick Sites For Me&quot; button.    Enter an amino acid interval and GeneDesign will select sites on that interval	using the criteria you defined in the last screen. The program's choices will be presented	to you using the same screen and you will have a chance to edit them. You can change the amino	acid interval and have the program reconsider as many times as you like.</p>  <p>Manual method: You can make your own landmark selections. Each pull down menu    is anchored by a line and a dot between a pair of amino acids; these only    appear where a silent mutation is possible. The pull down menus are populated    with lists of enzymes that can be introduced at those positions. A black    line indicates that none of the enzymes in the list are absent from the vector    you selected (or present in the list of sites you created). A red line indicates that    there are one or more enzymes in the list that are absent from the vector. Small warning: the program does	not currently keep a list of your choices during manual selection, to help you keep track of which enzymes	have been used already.  We are working on this.</p>  <p>If you have the computer select the sites it will use blue lines to indicate    its selections.  If you like you may edit the selection of sites after the computer has made its selections, 	for example, if you really want a unique site at a specific position, or if you require a XhoI site somewhere	in your gene.  The next screen will warn you of any difficulties in landmark insertion.</p>  <div id="imgrt"><img src="../img/ssi3.gif" alt="Restriction Site Addition Module" width="300" height="152"><br>    Restriction Site Addition Module - Summary of Sites</div>      <p>Once you are satisfied with the sites that have been selected you can        click the &quot;Continue to Summary&quot; button. The next screen will        provide you with a detailed list of the sites chosen for silent insertion.        The first column        is the site name. The next column is the recognition sequence, followed        by an asterisk if the enzyme exhibits star activity. The third column        is the nucleotide position of the site. The fourth column        defines the end type the enzyme will leave, 5', 3', or blunt. The fifth        column is the incubation temperature of the enzyme. The sixth and seventh        columns provide a general guideline to methylation sensitivity. The eighth        column is vendor availability and the ninth      column is the cost per unit in '04 dollars.</p>      <p>At the far left, there is a checkbox next to each enzyme name. If you        used automatic site selection and are unhappy with one or more of the        enzymes chosen, you can check any number of them and hit the &quot;Reconsider&quot; button.        Automatic selection will be run again and the enzymes you select will      be excluded from consideration.</p>      <p>Your new nucleotide sequence is at the top of this last screen.  It would be wise to double check it	   in your favorite sequence analysis program to be sure it produces          the correct amino acid sequence at this stage!<br><br>          You can now take this sequence to another 		  module.  To follow the &quot;Design a Gene&quot; path, click &quot;<a href="dgolides.html">Oligo		  Design</a>&quot;.    </p></div>	<div id="footer">	<a href="dgssi1.html" class="headli2">Previous</a>&nbsp;&nbsp;&nbsp;<a href="dgolides.html" class="headli">Next</a>	</div></div></body></html>