<html><head><title>GeneDesign Manual: Site Insertion (1)</title><meta http-equiv="Content-Type" content="text/html; charset=iso-8859-1"><link href="../acss/mn.css" rel="stylesheet" type="text/css"></head><body><div id="bigbox">	<div id="header">		<a href="../index.html"><img src="../img/gdlogobanner.gif" alt="GeneDesign"></a>  <a class="headli">Manual</a>	</div>	<div id="navbox">		<ul id="navlist">			<li><a href="about.html">About GeneDesign</a></li>			<li id="active"><a href="gendes.html" id="current">Design a Gene</a></li>			<ul id="subnavlist">				<li><a href="dgrevtrans.html">Reverse Translation</a></li>				<li id="active"><a href="dgssi1.html" id="current">Site Insertion (1)</a></li>				<li><a href="dgssi2.html">Site Insertion (2)</a></li>				<li><a href="dgolides.html">Oligo Design</a></li>			</ul>			<li><a href="mods.html">GeneDesign Modules</a></li>			<ul id="subnavlist">				<li><a href="codjug.html">Codon Juggling</a></li>				<li><a href="revtrans.html">Reverse Translation</a></li>				<li><a href="ssi.html">Silent Site Insertion</a></li>				<li><a href="ssr.html">Silent Site Removal</a></li>				<li><a href="enzcho.html">Enzyme Choosing</a></li>				<li><a href="seqana.html">Sequence Analysis</a></li>				<li><a href="olides.html">Oligo Design</a></li>				<li><a href="veccho.html">Vector Choosing</a></li>				<li><a href="shortr.html">Short Sequence Removal</a></li>				<li><a href="randna.html">Random DNA Sequences</a></li>			</ul>			<li><a href="errrep.html">Report an Error</a></li>			<li><a href="sugcom.html">Suggestions &amp; Comments</a></li>			<li><a href="enzymes.html">Enzyme List</a></li>			<li><a href="changes.html">Version History</a></li>		</ul>	</div>	<div id="content">  <strong>Design a Gene: Silent Site Insertion</strong><br>  <br>  The second step on the Design a Gene path is the Silent Site Insertion module.   This is where you will choose the vector you plan to use to carry your synthetic  gene.  Alternatively,  you  can   define a list of restriction enzymes to be considered for insertion.  Vector   choice is an extremely important step in the process. The purpose is to instruct   the program on what restriction sites you DO NOT WANT to use as special &quot;landmark   restriction sites&quot; in your synthetic sequence, namely those already present   in the vector. If you choose a very large vector with many sites in it, the   program will not have any sites left to play with. Therefore, if you have   a choice, use the smallest vector available to you. In practice pUC18/19 work   well; derivatives lacking the multiple cloning site work even better.<br>  <br>  Your input amino acid and first-pass (optimized) nucleotide sequences will  appear at the top of the screen.<br>  <div id="imgrt">  <img src="../img/ssi1.gif" alt="Silent Site Insertion Module" width="300" height="201"><br>  The Silent Site Insertion Module - Vector Selection</div>  <p>Next you must choose the sites to be inserted. If you specify a vector    sequence, all sites in the vector are automatically removed from consideration    as landmark restriction sites by	the program.  There    are three ways to choose eligible sites. You may <strong>(A)</strong> simply    select the sites from a list. Click on the site name in the left column    and then the arrow    button at the bottom of the column to move that site to the right hand column.    Repeat the process on the right side to remove it. Only sites in the right    column will be considered. Or you may <strong>(B)</strong> select the name    of the vector you will be using from a pull-down menu. GeneDesign has the    sequences of these vectors and will    determine which enzymes are absent from this vector automatically    for you. Lastly, you may <strong>(C)</strong> provide    a vector sequence and a name for that vector. GeneDesign will parse the sequence    and determine which enzymes are absent from your vector.</p>  <p>You may in addition specify a list of enzymes that must not appear in the    final sequence. <strong>(D)</strong> Click on the site name in the left column    and then the arrow button at the bottom of the column to move that site to    the right hand column.    Repeat the process on the right side to remove it. Only sites in the right    column will be processed. The sequence will be searched for these sites and    if they are found in your first-pass synthetic sequence, they will be removed. These    sites will not be considered for landmark restriction sites.<br>      <br>      At <strong>(E)</strong> you can define the ranking criteria for sites to      be inserted. We have found the default criteria to be more than adequate for      most purposes. For more information about ranking enzymes, see <a href="enzcho.html">Enzyme      Choosing</a>.<br>      <br>      Once      you have selected the vector and/or the sites you can click &quot;Next      Step&quot;      to move on to <a href="dgssi2.html">the next step in site        selection</a>. </p></div>	<div id="footer">	<a href="dgrevtrans.html" class="headli2">Previous</a>&nbsp;&nbsp;&nbsp;<a href="dgssi2.html" class="headli">Next</a>	</div></div></body></html>