<html><head><title>GeneDesign Manual: Building Block Design</title><meta http-equiv="Content-Type" content="text/html; charset=iso-8859-1"><link href="../acss/mn.css" rel="stylesheet" type="text/css"></head><body><div id="bigbox">	<div id="header">		<a href="../index.html"><img src="../img/gdlogobanner.gif" alt="GeneDesign"></a>  <a class="headli">Manual</a>	</div>	<div id="navbox">		<ul id="navlist">			<li><a href="about.html">About GeneDesign</a></li>			<li><a href="gendes.html">Design a Gene</a></li>			<ul id="subnavlist">				<li><a href="dgrevtrans.html">Reverse Translation</a></li>				<li><a href="dgssi1.html">Site Addition (1)</a></li>				<li><a href="dgssi2.html">Site Addition (2)</a></li>				<li><a href="dgolides.html">Building Block Design</a></li>			</ul>			<li id="active"><a href="mods.html" id="current">GeneDesign Modules</a></li>			<ul id="subnavlist">				<li><a href="codjug.html">Codon Juggling</a></li>				<li><a href="revtrans.html">Reverse Translation</a></li>				<li><a href="codbigr.html">Codon Bias Graphing</a></li>				<li><a href="rscugen.html">RSCU Value Generation</a></li>				<li><a href="ssi.html">Restriction Site Addition</a></li>				<li><a href="ssr.html">Restriction Site Subtraction</a></li>				<li><a href="enzcho.html">Enzyme Filter</a></li>				<li><a href="seqana.html">Sequence Analysis</a></li>				<li id="active"><a href="olides.html" id="current">Building Block Design</a></li>				<li><a href="veccho.html">Vector Choosing</a></li>				<li><a href="shortr.html">Short Sequence Subtraction</a></li>				<li><a href="randna.html">Random DNA Sequences</a></li>			</ul>			<li><a href="errrep.html">Report an Error</a></li>			<li><a href="sugcom.html">Suggestions &amp; Comments</a></li>			<li><a href="enzymes.html">Enzyme List</a></li>			<li><a href="changes.html">Version History</a></li>		</ul>	</div>	<div id="content">  <strong>GeneDesign Modules: Building Block Design</strong><br>  <br>  GeneDesign's <a href="../../cgi-bin/gd/gdOliDes.cgi">Building Block Design</a> module  allows you to generate a list of oligonucleotides from a nucleotide sequence.  The default settings produce 60bp oligos with ~20bp overlaps. These values  work well for us with yeast  and mammalian sequences which are ~40% GC. You may want to play with these  values if your sequences are outside of this range. GeneDesign will break the  sequence into roughly 500bp chunks on unique restriction sites (restriction  sites must be non-blunt, non-1bp overhang cutters that have cleavage sites  internal to the recognition sequence) and then break those chunks into an even  number of 60bp oligos. The oligos are  then  adjusted  for  even lengths  and equal melting temperatures.<br>  <div id="imgrt"><img src="../img/olides.gif" alt="Building Block Design module" width="300" height="138"><br>    The Oligo Selection Module - A Sample</div>  <p><br>    Every 500bp chunk is displayed on the oligo summary page.  For each chunk, the order number, length in bp, average 	Tm of overlaps, and 3' unique site appear in that order in the vertical bar at left.    The chunk sequence and the sequences of oligos are printed for visual confirmation of appropriate alignment at the 	top of each chunk.  Information for each oligo is contained in rows under the line up.  This information includes 	oligo length, start and stop coordinates, and sense.  It also includes 	data on the length and melting temperatures of the 5' and 3' overlaps. The 5' to 3' sequence of each oligo is printed at the far right.<br><br>	Melting temperature is calculated in three	ways. The first column uses the formula from <a href="http://www.ncbi.nlm.nih.gov/entrez/query.fcgi?cmd=Retrieve&db=pubmed&dopt=Abstract&list_uids=2725322&query_hl=9">Baldino	et al</a>. This is the same	formula	used by Strider. The second column uses	a derivative	of that formula, as used in <a href="http://frodo.wi.mit.edu/cgi-bin/primer3/primer3_www_help.cgi#PRIMER_PRODUCT_TM">Primer3</a>.	The third column uses the nearest neighbor (<a href="http://www.ncbi.nlm.nih.gov/entrez/query.fcgi?cmd=Retrieve&db=PubMed&dopt=Citation&list_uids=4427357">Borer	et al</a>) parameters for DNA/DNA	duplexes as defined in Sugimoto et al.	These parameters are used in the formula below, derived from <a href="http://www.ncbi.nlm.nih.gov/entrez/query.fcgi?cmd=Retrieve&db=pubmed&dopt=Abstract&list_uids=2243783&query_hl=12">Rychlik	et al</a>, where &Delta;H	is	enthalpy, &Delta;S	is entropy, R is the molar gas constant, (16.6*log[Na+]) is the salt correction,	and 3.4 kcal/mol is the activation energy required to go from single to double	stranded DNA.      </p>  <div align="center"><img src="../img/for3.gif" alt="Nearest Neighbor Melting Temperature Formula" width="482" height="69">  </div>  <br>Your oligos may be longer or shorter than expected because of optimizations for length and melting temperature.<br><br>If there are no further adjustments to be made, you can export these oligos to a delineated text file for ordering.    If you select a designation string   for the oligos and a delineator, you can use the &quot;Order Form&quot; button to kick out a list that can easily be opened in Excel or sent   to the oligo producing company of your choice.  The oligos will be numbered from first chunk to last, using the designation string.   For example, if your designation string is &quot;syngene&quot;, the first oligo will be called &quot;syngene01&quot; 	and the next will be &quot;syngene02&quot;, etc.  </div>	<div id="footer">	<a href="seqana.html" class="headli2">Previous</a>&nbsp;&nbsp;&nbsp;<a href="errrep.html" class="headli">Next</a>	</div></div></body></html>