<html><head><title>GeneDesign Manual: Silent Site Insertion</title><meta http-equiv="Content-Type" content="text/html; charset=iso-8859-1"><link href="../acss/mn.css" rel="stylesheet" type="text/css"></head><body><div id="bigbox">	<div id="header">		<a href="../index.html"><img src="../img/gdlogobanner.gif" alt="GeneDesign"></a>  <a class="headli">Manual</a>	</div>	<div id="navbox">		<ul id="navlist">			<li><a href="about.html">About GeneDesign</a></li>			<li><a href="gendes.html">Design a Gene</a></li>			<ul id="subnavlist">				<li><a href="dgrevtrans.html">Reverse Translation</a></li>				<li><a href="dgssi1.html">Site Insertion (1)</a></li>				<li><a href="dgssi2.html">Site Insertion (2)</a></li>				<li><a href="dgolides.html">Oligo Design</a></li>			</ul>			<li id="active"><a href="mods.html" id="current">GeneDesign Modules</a></li>			<ul id="subnavlist">				<li><a href="codjug.html">Codon Juggling</a></li>				<li><a href="revtrans.html">Reverse Translation</a></li>				<li id="active"><a href="ssi.html" id="current">Silent Site Insertion</a></li>				<li><a href="ssr.html">Silent Site Removal</a></li>				<li><a href="enzcho.html">Enzyme Choosing</a></li>				<li><a href="seqana.html">Sequence Analysis</a></li>				<li><a href="olides.html">Oligo Design</a></li>				<li><a href="veccho.html">Vector Choosing</a></li>				<li><a href="shortr.html">Short Sequence Removal</a></li>				<li><a href="randna.html">Random DNA Sequences</a></li>			</ul>			<li><a href="errrep.html">Report an Error</a></li>			<li><a href="sugcom.html">Suggestions &amp; Comments</a></li>			<li><a href="enzymes.html">Enzyme List</a></li>			<li><a href="changes.html">Version History</a></li>		</ul>	</div>	<div id="content">  <strong>GeneDesign Modules: Silent Site Insertion</strong><br>  <br>  GeneDesign's <a href="../../cgi-bin/gd/gdSSIns.cgi">Silent Site Insertion</a> module allows you to process a nucleotide sequence for modular  mutagenesis. You can choose the vector you plan to use to carry your synthetic  gene, or you can define a list of restriction enzymes to be considered for insertion.  Vector   choice is an extremely important step in the process. The purpose is to instruct   the program on what restriction sites you DO NOT WANT to use as special &quot;landmark   restriction sites&quot; in your synthetic sequence, namely those already present   in the vector. If you choose a very large vector with many sites in it, the   program will not have any sites left to play with. Therefore, if you have   a choice, use the smallest vector available to you. In practice pUC18/19 work   well; derivatives lacking the multiple cloning site work even better.<br>  <div id="imgrt">  <img src="../img/ssi1.gif" alt="Silent Site Insertion Module" width="300" height="201"><br>  The Silent Site Insertion Module - Vector Selection</div>  <p>Next you must choose the sites to be inserted. If you specify a vector    sequence, all sites in the vector are automatically removed from consideration    as landmark restriction sites by	the program.  There    are three ways to choose eligible sites. You may <strong>(A)</strong> simply    select the sites from a list. Click on the site name in the left column    and then the arrow    button at the bottom of the column to move that site to the right hand column.    Repeat the process on the right side to remove it. Only sites in the right    column will be considered. Or you may <strong>(B)</strong> select the name    of the vector you will be using from a pull-down menu. GeneDesign has the    sequences of these vectors and will    determine which enzymes are absent from this vector automatically    for you. Lastly, you may <strong>(C)</strong> provide    a vector sequence and a name for that vector. GeneDesign will parse the sequence    and determine which enzymes are absent from your vector.</p>  <p>You may in addition specify a list of enzymes that must not appear in the    final sequence. <strong>(D)</strong> Click on the site name in the left column    and then the arrow button at the bottom of the column to move that site to    the right hand column.    Repeat the process on the right side to remove it. Only sites in the right    column will be processed. The sequence will be searched for these sites and    if they are found in your first-pass synthetic sequence, they will be removed. These    sites will not be considered for landmark restriction sites.<br>      <br>      At <strong>(E)</strong> you can define the ranking criteria for sites to      be inserted. We have found the default criteria to be more than adequate for      most purposes. For more information about ranking enzymes, see <a href="enzcho.html">Enzyme      Choosing</a>.<br>      <br>      Once      you have selected the vector and/or the sites you can click &quot;Next      Step&quot;      to move on to the next step in site        selection. </p>  The second step of Silent Site Insertion involves the actual   selection of landmark restriction sites. Your protein sequence is displayed with an asterisk under every  tenth amino acid.  Anchored pulldown menus contain all pf the possible silent, unique insertions.<br>  <br><div id="imgrt">  <img src="../img/ssi2.gif" alt="Silent Site Insertion Module" width="300" height="251"><br>  The Silent Site Insertion Module - Landmark Selection</div>  <p>    Easiest method: You can have the computer select sites for you with the &quot;Pick    Sites For    Me&quot; button.    Enter an amino acid interval and GeneDesign will select sites on that interval	using the criteria you defined in the previous screen. The program's choices    will be presented to you using the same screen and you will have a chance    to edit them. You can change the amino	acid interval and have the program reconsider as many times as you like.</p>  <p>Manual method: You can make your own landmark selections. Each pulldown menu    is anchored by a line and a dot between a pair of amino acids; these only    appear where a silent mutation is possible. The pulldown menus are populated    with lists of enzymes that can be introduced at those positions. A black    line indicates that none of the enzymes in the list are absent from the vector    you selected (or present in the list of sites you created). A red line indicates that    there are one or more enzymes in the list that are absent from the vector.    </p>  <p>If you have the computer select the sites it will use blue lines to indicate    its selections.</p>  <div id="imgrt"><img src="../img/ssi3.gif" alt="Silent Site Insertion Module" width="300" height="152"><br>    Silent Site Insertion Module - Summary of Sites</div>      <p>Once you are satisfied with the sites that have been selected you can        click the &quot;Continue to Summary&quot; button. The next screen will        provide you with a detailed list of the sites chosen for silent insertion.        The first column        is the site name. The next column is the recognition sequence, followed        by an asterisk if the enzyme exhibits star activity. The third column        is the nucleotide position of the site. The fourth column        defines the end type the enzyme will leave, 5', 3', or blunt. The fifth        column is the incubation temperature of the enzyme. The sixth and seventh        columns provide a general guideline to methylation sensitivity. The eighth        column is vendor availability and the ninth      column is the cost per unit in '04 dollars.</p>      <p>At the far left, there is a checkbox next to each enzyme name. If you        used automatic site selection and are unhappy with one or more of the        enzymes chosen, you can check any number of them and hit the &quot;Reconsider&quot;        button. Automatic selection will be run again and the enzymes you select        will be excluded from consideration.</p>      <p>Your new nucleotide sequence is at the top of this last screen.  It would be wise to double check it	   in your favorite sequence analysis program to be sure it produces          the correct amino acid sequence at this stage!<br><br>          You can now take this sequence to another module.    </p>  </div>	<div id="footer">	<a href="revtrans.html" class="headli2">Previous</a>&nbsp;&nbsp;&nbsp;<a href="ssr.html" class="headli">Next</a>	</div></div></body></html>